milli/src/search/criteria/mod.rs

```diff
@@ -111,6 +118,68 @@ pub struct CriteriaBuilder<'t> {
     words_prefixes_fst: fst::Set<Cow<'t, [u8]>>,
 }
 
+/// Return the docids for the following word pairs and proximities using [`Context::word_pair_proximity_docids`].
+/// * `left, right, prox`   (leftward proximity)
+/// * `right, left, prox-1` (rightward proximity)
+///
+/// ## Example
+/// For a document with the text `the good fox eats the apple`, we have:
+/// * `rightward_proximity(the, eats) = 3`
+/// * `leftward_proximity(eats, the) = 1`
+///
+/// So both the expressions `word_pair_overall_proximity_docids(ctx, the, eats, 3)`
+/// and `word_pair_overall_proximity_docids(ctx, the, eats, 2)` would return a bitmap containing
+/// the id of this document.
+fn word_pair_overall_proximity_docids(
+    ctx: &dyn Context,
+    left: &str,
+    right: &str,
+    prox: u8,
+) -> heed::Result<Option<RoaringBitmap>> {
+    let rightward = ctx.word_pair_proximity_docids(left, right, prox)?;
+    let leftward =
+        if prox > 1 { ctx.word_pair_proximity_docids(right, left, prox - 1)? } else { None };
+    if let Some(mut all) = rightward {
+        if let Some(leftward) = leftward {
+            all |= leftward;
+        }
+        Ok(Some(all))
+    } else {
+        Ok(leftward)
+    }
+}
+
+/// This function works identically to [`word_pair_overall_proximity_docids`] except that the
+/// right word is replaced by a prefix string.
+///
+/// It will return None if no documents were found or if the prefix does not exist in the
+/// `word_prefix_pair_proximity_docids` database.
+fn word_prefix_pair_overall_proximity_docids(
+    ctx: &dyn Context,
+    left: &str,
+    prefix: &str,
+    proximity: u8,
+) -> heed::Result<Option<RoaringBitmap>> {
+    // We retrieve the docids for the original and swapped word pairs:
+    // A: word1 prefix2 proximity
+    // B: prefix2 word1 proximity-1
+    let rightward = ctx.word_prefix_pair_proximity_docids(left, prefix, proximity)?;
+
+    let leftward = if proximity > 1 {
+        ctx.prefix_word_pair_proximity_docids(prefix, left, proximity - 1)?
+    } else {
+        None
+    };
+    if let Some(mut all) = rightward {
+        if let Some(leftward) = leftward {
+            all |= leftward;
+        }
+        Ok(Some(all))
+    } else {
+        Ok(leftward)
+    }
```

Comment:
Is it possible to do something like:
```suggestion
    match rightward {
        Some(all) => Ok(leftward.map_or_else(all, |leftward| leftward | all)),
        None => Ok(leftward),
    }
```

Reply:
It is less efficient since we create a new bitmap instead of doing the operation in-place :/